internal/pkg/manifest/backend_svc.go

```diff
@@ -78,6 +78,21 @@ func (s *BackendService) MarshalBinary() ([]byte, error) {
 	return content.Bytes(), nil
 }
 
+// RequiredEnvironmentFeatures returns environment features that are required for this manifest.
+func (s *BackendService) RequiredEnvironmentFeatures() []string {
+	var features []string
+	if !s.RoutingRule.isEmpty() && !s.RoutingRule.Disabled() {
+		features = append(features, template.InternalALBFeatureName)
+	}
+	if aws.StringValue((*string)(s.Network.VPC.Placement.PlacementString)) == string(PrivateSubnetPlacement) {
+		features = append(features, template.NATFeatureName)
+	}
+	if efsFeatureRequired(s.Storage) {
+		features = append(features, template.EFSFeatureName)
+	}
+	return features
+}
```

Comment:
The concern was that the same configuration may mean different things for each workload type.

For example, an empty `http:` means ALB enabled for an LBWS and the opposite to a backend service. That was why I was letting the workload struct to decide what required features they want, instead of delegating to it to the subfields.

But yeah I think perhaps common logic can be delegated like you suggested. Will update!